File: Python/treehopper/libraries/sensors/inertial/Adxl345.py
from treehopper.api import I2C
from treehopper.libraries import SMBusDevice
from treehopper.libraries.sensors.inertial.adxl345_registers import Adxl345Registers
from treehopper.libraries.sensors.inertial import Accelerometer


class Adxl345(Accelerometer):
    """Analog Devices ADXL-345 three-axis I2C accelerometer

    Example:
        >>> from treehopper.api import *
        >>> from treehopper.libraries.sensors.inertial import Adxl345

        >>> board = find_boards()[0]
        >>> board.connect()
        >>> imu = Adxl345(alt_address=True, i2c=board.i2c)
        >>> while board.connected:
        >>>     print(imu.accelerometer)

    """
    def __init__(self, i2c: I2C, alt_address=False, rate=100):
        super().__init__()
        self._dev = SMBusDevice((0x53 if alt_address else 0x1D), i2c, rate)
        self._registers = Adxl345Registers(self._dev)
        self._registers.powerCtl.sleep = 0
        self._registers.powerCtl.measure = 1
        self._registers.dataFormat.range = 0x03
        self._registers.dataFormat.fullRes = 1
        self._registers.writeRange(self._registers.powerCtl, self._registers.dataFormat)

    def update(self):
        self._registers.readRange(self._registers.dataX, self._registers.dataZ)
        self._accelerometer = [self._registers.dataX.value / 255,
                              self._registers.dataY.value / 255,
                              self._registers.dataZ.value / 255]

File: C++/API/src/libraries/Sensors/Inertial/Adxl345.cpp
#include <Libraries/Sensors/Inertial/Adxl345.h>

namespace Treehopper {
    namespace Libraries {
        namespace Sensors {
            namespace Inertial {
                Adxl345::Adxl345(Treehopper::I2c &i2c, bool altAddress, int rate)
                        : _dev(!altAddress ? 0x53 : 0x1D, i2c, rate), registers(_dev) {
                    registers.powerCtl.sleep = 0;
                    registers.powerCtl.measure = 1;
                    registers.dataFormat.range = 0x03;
                    registers.dataFormat.fullRes = 1;
                    registers.writeRange(registers.powerCtl, registers.dataFormat);
                }

                void Adxl345::update() {
                    registers.readRange(registers.dataX, registers.dataZ);

                    _accelerometer.x = registers.dataX.value / 255.0f;
                    _accelerometer.y = registers.dataY.value / 255.0f;
                    _accelerometer.z = registers.dataZ.value / 255.0f;
                }
            }
        }
    }
}
